File: BFSandDFS.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Graph {
    int V;
    vector<vector<int>> adjList;

public:
    Graph(int vertices) {
        V = vertices;
        adjList.resize(V);
    }

    void addEdge(int u, int v) {
        adjList[u].push_back(v);
        adjList[v].push_back(u);
    }

    void dfsRecursive(int node, vector<bool>& visited) {
        visited[node] = true;
        cout << node << " ";

        for (int neighbor : adjList[node]) {
            if (!visited[neighbor]) {
                dfsRecursive(neighbor, visited);
            }
        }
    }

    void dfs(int start) {
        vector<bool> visited(V, false);
        dfsRecursive(start, visited);
    }

    void bfs(int start) {
        vector<bool> visited(V, false);
        queue<int> q;
        q.push(start);
        visited[start] = true;

        while (!q.empty()) {
            int node = q.front();
            q.pop();
            cout << node << " ";

            for (int neighbor : adjList[node]) {
                if (!visited[neighbor]) {
                    q.push(neighbor);
                    visited[neighbor] = true;
                }
            }
        }
    }
};

int main() {
    Graph g(5);
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 2);
    g.addEdge(2, 3);
    g.addEdge(3, 4);

    cout << "Depth-First Search (DFS): ";
    g.dfs(0);
    cout << endl;

    cout << "Breadth-First Search (BFS): ";
    g.bfs(0);
    cout << endl;

    return 0;
}

File: BFSandDFS.py
from collections import deque

graph = [[] for _ in range(5)]

graph[0].extend([1, 2])
graph[1].append(2)
graph[2].append(3)
graph[3].append(4)
graph[2].append(0)
graph[1].append(0)
graph[3].append(2)
graph[4].append(3)

def dfs(node, visited):
    visited[node] = True
    print(node, end=" ")
    for neighbor in graph[node]:
        if not visited[neighbor]:
            dfs(neighbor, visited)

def bfs(start):
    visited = [False] * len(graph)
    queue = deque([start])
    visited[start] = True

    while queue:
        node = queue.popleft()
        print(node, end=" ")
        for neighbor in graph[node]:
            if not visited[neighbor]:
                queue.append(neighbor)
                visited[neighbor] = True

print("DFS:", end=" ")
visited = [False] * len(graph)
dfs(0, visited)

print("\nBFS:", end=" ")
bfs(0)
